src/pyhf/tensor/common.py

```diff
@@ -0,0 +1,38 @@
+from .. import default_backend, get_backend
+from .. import events
+
+
+class TensorViewer(object):
+    def __init__(self, indices):
+        self.partition_indices = indices
+        a = default_backend.astensor(
+            default_backend.concatenate(self.partition_indices), dtype='int'
+        )
+        self._sorted_indices = default_backend.tolist(a.argsort())
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        tensorlib, _ = get_backend()
+        self.sorted_indices = tensorlib.astensor(self._sorted_indices)
+
+    def stitch(self, data):
+        tensorlib, _ = get_backend()
+        assert len(self.partition_indices) == len(data)
+
+        data = tensorlib.concatenate(data, axis=-1)
+        data = tensorlib.einsum('...j->j...', data)
+        stitched = tensorlib.gather(data, tensorlib.astensor(self.sorted_indices))
+        stitched = tensorlib.einsum('...j->j...', stitched)
```

Comment:
why are we transposing twice?

Reply:
we need to split the last axis, so we move an axis up front in order to be able split it with inndices and then once we're done move the axis back to the last position